File: wind_speed/scripts/wind_speed_foucaut.py
"""
animate_foucaut_wind.py

10m wind speed climatology animation in the Foucaut projection, with mean
wind vectors overlaid.

Same Foucaut projection class as animate_foucaut_2m_temp.py / 
animate_foucaut_SLP.py. Longitude is normalized to [-180, 180) and sorted
for the WHOLE dataset up front (fixes the pcolormesh antimeridian
distortion seen on the temp version) -- this makes the original script's
quiver-only local longitude re-sort redundant, so that's been removed
rather than left in as dead logic.

Since Foucaut has a working inverse transform (unlike Nicolosi), Cartopy's
ax.quiver(transform=ccrs.PlateCarree()) handles wind vector rotation
automatically -- no manual Jacobian rotation needed here, unlike the
Nicolosi wind script.

NOTE: still not execution-tested end to end (ongoing tool access issue).
The Foucaut class and longitude fix carry over from the temp/SLP versions,
which the user confirmed working; wind-specific changes (variables,
colormap, quiver call) mirror animate_wind.py but haven't been run.
"""

import xarray as xr
import numpy as np
import matplotlib
import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import matplotlib.ticker
import cartopy.crs as ccrs
import cartopy.feature as cfeature
import shapely.geometry as sgeom
import imageio.v2 as imageio
from pathlib import Path

matplotlib.rcParams['font.family'] = 'Arial'

# ── Paths ────────────────────────────────────────────────────────────────────
DATA_FILE = Path('/Users/waverleymoody/Downloads/climate_data_by_variable/wind_climatology.nc')
OUTPUT_DIR = Path('/Users/waverleymoody/Downloads/wind_animation')
OUTPUT_DIR.mkdir(exist_ok=True)


# ── Custom Foucaut projection ──────────────────────────────────────────────────
class Foucaut(ccrs.Projection):
    """PROJ's plain `fouc` operation (NOT `fouc_s`/Foucaut Sinusoidal).
    Confirmed working via animate_foucaut_2m_temp.py."""

    def __init__(self):
        super().__init__(proj4_params=[('proj', 'fouc')])

        lats = np.linspace(-90, 90, 200)
        lons_left = np.full_like(lats, -180)
        lons_right = np.full_like(lats, 180)
        all_lons = np.concatenate([lons_left, lons_right[::-1]])
        all_lats = np.concatenate([lats, lats[::-1]])

        xyz = self.transform_points(ccrs.Geodetic(), all_lons, all_lats)
        x, y = xyz[:, 0], xyz[:, 1]
        valid = np.isfinite(x) & np.isfinite(y)
        x, y = x[valid], y[valid]

        self._boundary = sgeom.LinearRing(zip(x, y))
        self._x_limits = (x.min(), x.max())
        self._y_limits = (y.min(), y.max())

    @property
    def boundary(self):
        return self._boundary

    @property
    def x_limits(self):
        return self._x_limits

    @property
    def y_limits(self):
        return self._y_limits

    @property
    def threshold(self):
        return 1e4


PROJ_CRS = Foucaut()
PROJ_NAME = 'foucaut'

# ── Plot settings (identical to animate_wind.py) ──────────────────────────────
VMIN, VMAX = 0, 16
colors = [
    (0.00,  '#ffffff'),  # white        (0 m/s)
    (0.20,  '#d0e8f0'),  # light blue   (3 m/s)
    (0.35,  "#469a46"),  # light green        (5.5 m/s)
    (0.44,  '#ffff00'),  # yellow       (7 m/s)
    (0.50,  '#ff6600'),  # orange       (8 m/s)
    (0.60,  '#ff3300'),  # darker orange(9.5 m/s)
    (0.70,  '#ff0000'),  # light red    (11 m/s)
    (0.80,  '#cc0000'),  # red          (13 m/s)
    (0.90,  '#990000'),  # dark red     (14.5 m/s)
    (1.00,  '#660000'),  # darkest red  (16 m/s)
]
CMAP = mcolors.LinearSegmentedColormap.from_list('custom_wind', colors, N=35)

# ── Load pre-computed climatology ──────────
ds_clim = xr.open_dataset(DATA_FILE)

# Normalize longitude to [-180, 180) and sort ascending -- fixes the
# pcolormesh antimeridian distortion for the whole dataset up front,
# which also makes the quiver subsampling's local re-sort unnecessary.
lon_vals = ds_clim['longitude'].values
lon_norm = ((lon_vals + 180) % 360) - 180
ds_clim = ds_clim.assign_coords(longitude=lon_norm).sortby('longitude')

lats = ds_clim['latitude'].values
lons = ds_clim['longitude'].values
labels = ds_clim['frame_label'].values

frames_data = []
for i in range(ds_clim.sizes['frame']):
    mean_speed = ds_clim['wind_speed'].isel(frame=i).values
    frame_u = ds_clim['u10'].isel(frame=i).values
    frame_v = ds_clim['v10'].isel(frame=i).values
    label = str(labels[i])
    frames_data.append((label, mean_speed, frame_u, frame_v))
    print(f'Loaded: {label}')

# ── Generate frames + MP4 ──────────────────────────────────────────────────────
frames_dir = OUTPUT_DIR / PROJ_NAME / 'frames'
frames_dir.mkdir(parents=True, exist_ok=True)
frame_paths = []

for i, (label, mean_speed, frame_u, frame_v) in enumerate(frames_data):
    fig, ax = plt.subplots(figsize=(12, 6), subplot_kw={'projection': PROJ_CRS})

    im = ax.pcolormesh(
        lons, lats, mean_speed,
        vmin=VMIN, vmax=VMAX,
        cmap=CMAP,
        transform=ccrs.PlateCarree(),
        shading='auto'
    )

    # lons/lats are already normalized+sorted at the dataset level above,
    # so subsampling here needs no separate re-sort (unlike the original
    # script, where lons was still 0-360 at this point).
    step = 30
    lons_sub = lons[::step]
    lats_sub = lats[::step]
    u_sub = frame_u[::step, ::step]
    v_sub = frame_v[::step, ::step]
    lons_grid, lats_grid = np.meshgrid(lons_sub, lats_sub)

    # Foucaut has a working inverse transform (unlike Nicolosi), so
    # Cartopy handles wind vector rotation automatically here -- same as
    # it already does for the Robinson output.
    ax.quiver(
        lons_grid, lats_grid,
        u_sub, v_sub,
        transform=ccrs.PlateCarree(),
        color='black',
        scale=300,
        width=0.001,
        headwidth=3,
        headlength=3,
    )

    ax.add_feature(cfeature.COASTLINE, linewidth=1.0, edgecolor='black')
    ax.add_feature(cfeature.BORDERS, linewidth=0.8, edgecolor='black')
    ax.add_feature(cfeature.STATES, linewidth=0.5, edgecolor='black')
    ax.set_global()

    gl = ax.gridlines(draw_labels=False, linewidth=0.3, color='gray', alpha=0.5)
    gl.xlocator = matplotlib.ticker.FixedLocator([-180, -90, 0, 90, 180])
    gl.ylocator = matplotlib.ticker.FixedLocator([-90, -45, 0, 45, 90])

    # No longitude labels (removed per preference). Single pole labels,
    # since the whole pole is one point in this projection.
    x, y = PROJ_CRS.transform_point(0, 90, ccrs.PlateCarree())
    ax.annotate('90°N', xy=(x, y), xycoords='data',
                xytext=(0, 6), textcoords='offset points',
                ha='center', va='bottom', fontsize=9, annotation_clip=False)
    x, y = PROJ_CRS.transform_point(0, -90, ccrs.PlateCarree())
    ax.annotate('90°S', xy=(x, y), xycoords='data',
                xytext=(0, -6), textcoords='offset points',
                ha='center', va='top', fontsize=9, annotation_clip=False)

    # Latitude labels along the left edge, excluding the poles (handled above).
    for lat, lat_label in [(-45, '45°S'), (0, '0°'), (45, '45°N')]:
        x, y = PROJ_CRS.transform_point(-180, lat, ccrs.PlateCarree())
        ax.annotate(lat_label, xy=(x, y), xycoords='data',
                    xytext=(-8, 0), textcoords='offset points',
                    ha='right', va='center', fontsize=9, annotation_clip=False)

    cbar = plt.colorbar(im, ax=ax, orientation='horizontal',
                        pad=0.06, fraction=0.12, shrink=0.5,
                        aspect=20, extend='neither')
    cbar.set_label('Wind Speed at 10 m (m/s)', fontsize=11)
    cbar.ax.tick_params(labelsize=9)
    cbar.set_ticks(np.arange(0, 17, 2))
    cbar.set_ticklabels([str(t) for t in np.arange(0, 17, 2)])

    ax.text(0.0, 1.02, 'ERA-5 | Climate Reanalyzer',
            transform=ax.transAxes, fontsize=10,
            fontweight='bold', ha='left', va='bottom')
    ax.text(1.0, 1.02, f'{label}; 1979–2000 Weekly Mean',
            transform=ax.transAxes, fontsize=10,
            fontweight='bold', ha='right', va='bottom')

    fpath = frames_dir / f'frame_{i:03d}.png'
    plt.savefig(fpath, dpi=150, bbox_inches='tight')
    plt.close()
    frame_paths.append(str(fpath))
    print(f'Saved frame {i+1}/{len(frames_data)}')

video_name = 'wind_climatology_foucaut.mp4'
with imageio.get_writer(OUTPUT_DIR / video_name, fps=4, codec='libx264') as writer:
    for fp in frame_paths:
        writer.append_data(imageio.imread(fp))

print(f'Done! Animation saved to: {OUTPUT_DIR / video_name}')

File: wind_speed/scripts/wind_speed_foucaut.R
# SDSU Climate Informatics Lab
# San Diego State University
# by Waverley Moody
# Supervised by Distinguished Professor Samuel Shen
# R Code Version 1.0.0
#
# A reproduction of the University of Washington General Circulation
# Animations Library by Professor John Michael Wallace.
#
# Script: wind_speed_foucaut.R
# Description: Generates the 10m surface wind speed climatology animation from ERA5 reanalysis data (1979-2000), with mean wind vectors overlaid, rendered in the Foucaut projection.
#
# Note: For the Plate Carrée, Robinson, and Nicolosi projections, see the other scripts in the wind_speed scripts folder.

library(terra)
library(ncdf4)
library(rnaturalearth)
library(sf)
library(av)

# ── Paths ────────────────────────────────────────────────────────────────
data_file  <- "/Users/waverleymoody/Downloads/climate_data_by_variable/wind_climatology.nc"
output_dir <- "/Users/waverleymoody/Downloads/wind_animation_R"
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

proj_name  <- "foucaut"
target_crs <- "+proj=fouc +datum=WGS84 +units=m +no_defs"

frames_dir <- file.path(output_dir, proj_name, "frames")
dir.create(frames_dir, showWarnings = FALSE, recursive = TRUE)

# ── Plot settings (identical to animate_wind.py) ───────────────────────────
vmin <- 0
vmax <- 16

# Starting point from the Robinson/PlateCarree tuning session -- bumped up
# from 0.7 for longer arrow tails on Foucaut.
arrow_scale_deg <- 1.0
arrow_step      <- 35

# ── Custom colormap with uneven stop positions ──────────────────────────────
# Mirrors LinearSegmentedColormap.from_list with (position, color) tuples,
# since colorRampPalette() only supports evenly-spaced stops.
make_custom_palette <- function(colors, positions, n) {
  rgb_mat  <- t(col2rgb(colors)) / 255
  eval_pos <- seq(0, 1, length.out = n)
  r <- approx(positions, rgb_mat[, 1], xout = eval_pos)$y
  g <- approx(positions, rgb_mat[, 2], xout = eval_pos)$y
  b <- approx(positions, rgb_mat[, 3], xout = eval_pos)$y
  rgb(r, g, b)
}

wind_color_stops <- c(
  '#ffffff', # white         (0 m/s)
  '#d0e8f0', # light blue    (3 m/s)
  '#469a46', # light green   (5.5 m/s)
  '#ffff00', # yellow        (7 m/s)
  '#ff6600', # orange        (8 m/s)
  '#ff3300', # darker orange (9.5 m/s)
  '#ff0000', # light red     (11 m/s)
  '#cc0000', # red           (13 m/s)
  '#990000', # dark red      (14.5 m/s)
  '#660000'  # darkest red   (16 m/s)
)
wind_positions <- c(0.00, 0.20, 0.35, 0.44, 0.50, 0.60, 0.70, 0.80, 0.90, 1.00)

n_colors <- 35
wind_palette <- make_custom_palette(wind_color_stops, wind_positions, n_colors)
breaks <- seq(vmin, vmax, length.out = n_colors + 1)

# ── Natural Earth basemap layers ───────────────────────────────────────────
cat("Downloading/loading basemap layers...\n")
coastline <- ne_coastline(scale = 50, returnclass = "sf")
borders   <- ne_countries(scale = 50, returnclass = "sf")
states    <- ne_states(country = "United States of America", returnclass = "sf")

# Reproject once, not per frame
coastline_p <- st_transform(coastline, target_crs)
borders_p   <- st_transform(borders, target_crs)
states_p    <- st_transform(states, target_crs)

# ── Load pre-computed climatology ──────────────────────────────────────────
speed_stack <- rast(data_file, subds = "wind_speed")
u_stack     <- rast(data_file, subds = "u10")
v_stack     <- rast(data_file, subds = "v10")

# ERA5 is 0-360 -- fix all three consistently before reprojecting/subsampling
speed_stack <- rotate(speed_stack)
u_stack     <- rotate(u_stack)
v_stack     <- rotate(v_stack)
n_frames    <- nlyr(speed_stack)

nc <- nc_open(data_file)
frame_labels <- ncvar_get(nc, "frame_label")
nc_close(nc)

cat(sprintf("Loaded %d frames\n", n_frames))

# ── Helper: project a single lon/lat point into a target CRS ───────────────
proj_point <- function(lon, lat, target_crs) {
  pts <- vect(cbind(lon, lat), crs = "EPSG:4326")
  pts_proj <- project(pts, target_crs)
  crds(pts_proj)
}

# ── Helper: build a graticule line (meridian or parallel) in a target CRS ──
make_meridian <- function(lon, target_crs) {
  lats <- seq(-89.9, 89.9, length.out = 100)
  ln <- st_linestring(cbind(lon, lats))
  st_transform(st_sfc(ln, crs = "EPSG:4326"), target_crs)
}
make_parallel <- function(lat, target_crs) {
  lons <- seq(-180, 180, length.out = 200)
  ln <- st_linestring(cbind(lons, lat))
  st_transform(st_sfc(ln, crs = "EPSG:4326"), target_crs)
}

meridians <- lapply(c(-180, -90, 0, 90, 180), make_meridian, target_crs = target_crs)
parallels <- lapply(c(-90, -45, 0, 45, 90), make_parallel, target_crs = target_crs)

# ── Build clipping boundary (the true Foucaut diamond outline) ─────────────
# See 2m_temp_foucaut.R for the full explanation.
lats_b    <- seq(-90, 90, length.out = 200)
left_pts  <- vect(cbind(-180, lats_b), crs = "EPSG:4326")
right_pts <- vect(cbind(180, rev(lats_b)), crs = "EPSG:4326")
boundary_pts    <- rbind(project(left_pts, target_crs), project(right_pts, target_crs))
boundary_coords <- crds(boundary_pts)
boundary_coords <- rbind(boundary_coords, boundary_coords[1, ])  # close the ring
boundary_poly   <- vect(boundary_coords, type = "polygons", crs = target_crs)

# ── Arrow grid (lon/lat positions only -- frame-invariant, built once) ─────
col_idx  <- seq(1, ncol(u_stack), by = arrow_step)
row_idx  <- seq(1, nrow(u_stack), by = arrow_step)
lons_sub <- xFromCol(u_stack, col_idx)
lats_sub <- yFromRow(u_stack, row_idx)

idx_grid <- expand.grid(ri = seq_along(row_idx), ci = seq_along(col_idx))
grid_lon <- lons_sub[idx_grid$ci]
grid_lat <- lats_sub[idx_grid$ri]

# ── Generate frames + MP4 ───────────────────────────────────────────────────
frame_paths <- character(n_frames)

for (i in seq_len(n_frames)) {
  
  speed_field_p <- mask(project(speed_stack[[i]], target_crs), boundary_poly)
  
  # Per-frame u/v values at the fixed arrow grid positions
  u_mat   <- as.matrix(u_stack[[i]], wide = TRUE)
  v_mat   <- as.matrix(v_stack[[i]], wide = TRUE)
  grid_u  <- u_mat[cbind(row_idx[idx_grid$ri], col_idx[idx_grid$ci])]
  grid_v  <- v_mat[cbind(row_idx[idx_grid$ri], col_idx[idx_grid$ci])]
  
  end_lon <- grid_lon + grid_u * arrow_scale_deg
  end_lon <- ((end_lon + 180) %% 360) - 180   # dateline wrap
  end_lat <- pmin(pmax(grid_lat + grid_v * arrow_scale_deg, -90), 90)
  
  valid    <- is.finite(grid_u) & is.finite(grid_v)
  start_xy <- proj_point(grid_lon[valid], grid_lat[valid], target_crs)
  end_xy   <- proj_point(end_lon[valid], end_lat[valid], target_crs)
  
  # Outlier filter -- drops the handful of bad-projection points per frame
  # that produced stray horizontal-line artifacts on Robinson. This is the
  # fix that actually matters; arrow_scale_deg/arrow_step are cosmetic.
  map_ext       <- ext(speed_field_p)
  map_width     <- as.numeric(map_ext[2] - map_ext[1])
  max_arrow_len <- map_width * 0.03
  arrow_len     <- sqrt((end_xy[, 1] - start_xy[, 1])^2 + (end_xy[, 2] - start_xy[, 2])^2)
  keep          <- arrow_len < max_arrow_len
  
  label <- frame_labels[i]
  fpath <- file.path(frames_dir, sprintf("frame_%03d.png", i - 1))
  
  png(fpath, width = 1800, height = 980, res = 150)
  layout(matrix(1:2, nrow = 2), heights = c(5, 1.3))
  
  # ── Map panel ──
  par(mar = c(2, 3, 7.5, 2), bty = "n")
  plot(speed_field_p, col = wind_palette, breaks = breaks, range = c(vmin, vmax),
       axes = FALSE, legend = FALSE, main = "")
  
  plot(st_geometry(coastline_p), add = TRUE, lwd = 1.0, col = "black")
  plot(st_geometry(borders_p), add = TRUE, lwd = 0.8, border = "black")
  plot(st_geometry(states_p), add = TRUE, lwd = 0.5, border = "black")
  
  for (m in meridians) plot(m, add = TRUE, col = "gray", lwd = 0.3)
  for (p in parallels) plot(p, add = TRUE, col = "gray", lwd = 0.3)
  
  arrows(start_xy[keep, 1], start_xy[keep, 2], end_xy[keep, 1], end_xy[keep, 2],
         length = 0.03, angle = 20, code = 2, lwd = 0.6, col = "black")
  
  # Pole labels -- single points, not a longitude loop (same as temp/SLP)
  xy_n <- proj_point(0, 90, target_crs)
  text(xy_n[1], xy_n[2], labels = "90\u00b0N", pos = 3, cex = 0.95,
       xpd = NA, offset = 0.3)
  xy_s <- proj_point(0, -90, target_crs)
  text(xy_s[1], xy_s[2], labels = "90\u00b0S", pos = 1, cex = 0.95,
       xpd = NA, offset = 0.3)
  
  # Latitude labels along the left edge only. No longitude labels.
  lat_labels <- list(c(-45, "45\u00b0S"), c(0, "0\u00b0"), c(45, "45\u00b0N"))
  for (ll in lat_labels) {
    lat_val <- as.numeric(ll[1]); lab <- ll[2]
    xy <- proj_point(-180, lat_val, target_crs)
    text(xy[1], xy[2], labels = lab, pos = 2, cex = 0.95, xpd = NA, offset = 0.4)
  }
  
  mtext("ERA-5 | Climate Reanalyzer", side = 3, line = 6, adj = 0,
        cex = 1.0, font = 2)
  mtext(sprintf("%s; 1979\u20132000 Weekly Mean", label), side = 3, line = 6,
        adj = 1, cex = 1.0, font = 2)
  
  # ── Colorbar panel ──
  # No divider lines here -- animate_wind.py doesn't have them (unlike
  # animate_SLP.py), so none added to match.
  par(mar = c(4.5, 6, 1.2, 6), bty = "n")
  pad <- (vmax - vmin) * 0.25
  # x given as cell edges (breaks), not centers -- see SLP_foucaut.R for
  # why: centers let image() auto-extend the outer edges past vmin/vmax,
  # which poked color past the border rect.
  cbar_z <- matrix(seq_len(n_colors), ncol = 1)
  image(x = breaks, y = c(0.5, 1.5), z = cbar_z,
        col = wind_palette, breaks = seq(0.5, n_colors + 0.5, by = 1),
        axes = FALSE, xlab = "", ylab = "",
        xlim = c(vmin - pad, vmax + pad))
  usr <- par("usr")
  rect(vmin, usr[3], vmax, usr[4], border = "black", lwd = 1.2)
  axis(1, at = seq(vmin, vmax, 2), labels = seq(vmin, vmax, 2), cex.axis = 1.0)
  mtext("Wind Speed at 10 m (m/s)", side = 1, line = 3.0, cex = 1.0)
  
  dev.off()
  
  frame_paths[i] <- fpath
  cat(sprintf("Saved frame %d/%d\n", i, n_frames))
}

video_path <- file.path(output_dir, "wind_climatology_foucaut.mp4")
av_encode_video(frame_paths, output = video_path, framerate = 4, codec = "libx264")

cat(sprintf("Done! Animation saved to: %s\n", video_path))